src/endpoint.rs

```diff
@@ -20,154 +22,53 @@ use crate::{
-            }),
-        }
-    }
+pub(crate) type DynEndpoint<AppData> =
+    dyn (Fn(Context<AppData>) -> FutureObj<'static, Response>) + 'static + Send + Sync;
 
-    pub fn call(
-        &self,
-        data: Data,
-        req: Request,
-        params: Option<RouteMatch<'_>>,
-        store: &Store,
-    ) -> FutureObj<'static, Response> {
-        (self.endpoint)(data, req, params, store)
+impl<AppData, F: Send + Sync + 'static, Fut> Endpoint<AppData> for F
+where
+    F: Fn(Context<AppData>) -> Fut,
+    Fut: Future + Send + 'static,
+    Fut::Output: IntoResponse,
+{
+    type Fut = ResponseWrapper<Fut>;
+    fn call(&self, cx: Context<AppData>) -> Self::Fut {
+        ResponseWrapper { fut: (self)(cx) }
     }
 }
 
-/// A marker type used for the (phantom) `Kind` parameter in endpoints.
-#[doc(hidden)]
-pub struct Ty<T>(T);
-
-macro_rules! call_f {
-    ($head_ty:ty; ($f:ident, $head:ident); $($X:ident),*) => {
-        $f($head.clone(), $($X),*)
-    };
-    (($f:ident, $head:ident); $($X:ident),*) => {
-        $f($($X),*)
-    };
+/// The future retured by the endpoint implementation for `Fn` types.
+pub struct ResponseWrapper<F> {
+    fut: F,
 }
 
-macro_rules! end_point_impl_raw {
-    ($([$head:ty])* $($X:ident),*) => {
-        impl<T, Data, Fut, $($X),*> Endpoint<Data, (Ty<Fut>, $($head,)* $(Ty<$X>),*)> for T
-        where
-            T: Send + Sync + Clone + 'static + Fn($($head,)* $($X),*) -> Fut,
-            Data: Clone + Send + Sync + 'static,
-            Fut: Future + Send + 'static,
-            Fut::Output: IntoResponse,
-            $(
-                $X: Extract<Data>
-            ),*
-        {
-            type Fut = FutureObj<'static, Response>;
-
-            #[allow(unused_mut, non_snake_case)]
-            fn call(&self, mut data: Data, mut req: Request, params: Option<RouteMatch<'_>>, store: &Store) -> Self::Fut {
-                let f = self.clone();
-                $(let $X = $X::extract(&mut data, &mut req, &params, store);)*
-                FutureObj::new(Box::new(async move {
-                    let (parts, _) = req.into_parts();
-                    let head = Head::from(parts);
-                    $(let $X = match await!($X) {
-                        Ok(x) => x,
-                        Err(resp) => return resp,
-                    };)*
-                    let res = await!(call_f!($($head;)* (f, head); $($X),*));
+impl<F> Future for ResponseWrapper<F>
+where
+    F: Future,
+    F::Output: IntoResponse,
+{
+    type Output = Response;
 
-                    res.into_response()
-                }))
-            }
-        }
-    };
-}
-
-macro_rules! end_point_impl {
-    ($($X:ident),*) => {
-        end_point_impl_raw!([Head] $($X),*);
-        end_point_impl_raw!($($X),*);
+    fn poll(self: Pin<&mut Self>, waker: &std::task::Waker) -> std::task::Poll<Response> {
+        let inner = unsafe { self.map_unchecked_mut(|wrapper| &mut wrapper.fut) };
```

Comment:
Maybe a comment explaining why we need `unsafe` here?

Reply:
Better yet -- nixed the unsafe :)